File: app/src/main/java/com/sysu/edu/academic/PhysicalFitnessTestResultActivity.java
package com.sysu.edu.academic;

import android.os.Bundle;
import android.os.Handler;
import android.os.Message;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.RecyclerView;
import androidx.viewbinding.ViewBinding;

import com.google.android.material.tabs.TabLayoutMediator;
import com.sysu.edu.R;
import com.sysu.edu.api.HttpManager;
import com.sysu.edu.api.Params;
import com.sysu.edu.api.TargetUrl;
import com.sysu.edu.databinding.ActivityPagerBinding;
import com.sysu.edu.view.AdapterListener;
import com.sysu.edu.view.Pager2Adapter;
import com.sysu.edu.view.StaggeredFragment;

import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PhysicalFitnessTestResultActivity extends AppCompatActivity {

    Params params;
    HttpManager http;
    int position;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityPagerBinding binding = ActivityPagerBinding.inflate(getLayoutInflater());
        setContentView(binding.getRoot());
        binding.toolbar.setTitle(R.string.physical_fitness_test_result);
        binding.toolbar.setNavigationOnClickListener(_ -> supportFinishAfterTransition());
        params = new Params(this);
        params.setCallback(this::getResult);
        Pager2Adapter adp = new Pager2Adapter(this);
        StaggeredFragment page1 = new StaggeredFragment();
        StaggeredFragment page2 = new StaggeredFragment();
        StaggeredFragment page3 = new StaggeredFragment();
        adp.add(page1).add(page2).add(page3);
        binding.pager.setAdapter(adp);
        new TabLayoutMediator(binding.tabs, binding.pager, (tab, position) -> tab.setText(List.of("体测成绩", "体育积分", "游泳").get(position))).attach();
        Handler handler = new Handler(getMainLooper()) {
            @Override
            public void handleMessage(@NonNull Message msg) {
                super.handleMessage(msg);
                int num = msg.what;
                if (num == -1) params.toast(R.string.no_net_connected);
                else {
                    if (msg.getData().getInt("code") == 302 || Pattern.compile("window\\.location\\.href.+?\"/caslogin\"", Pattern.DOTALL).matcher((String) msg.obj).find()) {
                        params.toast(R.string.login_warning);
                        params.gotoLogin(TargetUrl.TICE);
                    } else switch (num) {
                        case 0, 1, 2 -> {
                            ArrayList<String> urls = new ArrayList<>();
                            Matcher matcher = Pattern.compile("<a class=\"weui-cell weui-cell_access\".+?</a>", Pattern.DOTALL).matcher((String) msg.obj);
                            StaggeredFragment page = (StaggeredFragment) adp.get(num);
                            while (matcher.find()) {
                                Matcher matcher1 = Pattern.compile("<div class=\"weui-cell__bd\".*?<p.*?>(.+?)</p>.*?<div class=\"weui-cell__ft.*?\">.+?<span.+?>(.+?)(&nbsp;)?</span>", Pattern.DOTALL).matcher(matcher.group());
                                if (matcher1.find())
                                    page.add(matcher1.group(1), R.drawable.calendar, List.of(new String[]{"总成绩", "总积分", "是否达标"}[msg.what]), List.of(Objects.requireNonNull(matcher1.group(2)).trim()));
                                Matcher matcher2 = Pattern.compile("<a class=\"weui-cell weui-cell_access\" href=\"(.+?)\">", Pattern.DOTALL).matcher(matcher.group());
                                if (matcher2.find()) urls.add(matcher2.group(1));
                            }
                            page.setListener(new AdapterListener() {
                                @Override
                                public void onBind(RecyclerView.Adapter<RecyclerView.ViewHolder> adapter, RecyclerView.ViewHolder holder, int position) {
                                    page.getStaggeredAdapter().getTwoColumnsAdapter(position).setListener(new AdapterListener() {
                                        @Override
                                        public void onBind(RecyclerView.Adapter<RecyclerView.ViewHolder> adapter, RecyclerView.ViewHolder holder, int p) {
                                            holder.itemView.setOnClickListener(_ -> {
                                                if (adapter.getItemCount() == 1 && num != 2) {
                                                    if (num == 0) getDetail(urls.get(position));
                                                    if (num == 1)
                                                        getCreditDetail(urls.get(position));
                                                    PhysicalFitnessTestResultActivity.this.position = position;
                                                }
                                            });
                                        }

                                        @Override
                                        public void onCreate(RecyclerView.Adapter<RecyclerView.ViewHolder> adapter, ViewBinding binding) {

                                        }
                                    });
                                }

                                @Override
                                public void onCreate(RecyclerView.Adapter<RecyclerView.ViewHolder> adapter, ViewBinding binding) {

                                }
                            });
                            if (msg.what == 0) getCredit();
                            if (msg.what == 1) getSwim();
                        }
                        case 3 -> {
                            Matcher matcher = Pattern.compile("<a class=\"weui-cell\">.*?<div class=\"weui-cell__bd\">(.+?)</div>.*?<div class=\"weui-cell__ft\">(.+?)</div>.*?</a>", Pattern.DOTALL).matcher((String) msg.obj);
                            while (matcher.find())
                                page1.getStaggeredAdapter().addRow(position, Objects.requireNonNull(matcher.group(1)).replace("</span>", "~").replaceAll("<.+?>", "").replaceAll("\\s", "").trim(), Objects.requireNonNull(matcher.group(2)).replace("</span>", "~").replaceAll("<.+?>", "").replace("%s", "").trim());
                        }
                        case 4 -> {
                            Matcher matcher = Pattern.compile("<a class=\"weui-cell\">.+?</a>", Pattern.DOTALL).matcher((String) msg.obj);
                            while (matcher.find()) {
                                Matcher matcher1 = Pattern.compile("class=\"left_side\">(.+?)</div>.+?<p>(.+?)</p></div>", Pattern.DOTALL).matcher(matcher.group());
                                Matcher creditMatcher = Pattern.compile("class=\"ticeImg\">(.+?)<", Pattern.DOTALL).matcher(matcher.group());
                                if (creditMatcher.find())
                                    page2.getStaggeredAdapter().addRow(position, "积分", Objects.requireNonNull(creditMatcher.group(1)).trim());
                                while (matcher1.find())
                                    page2.getStaggeredAdapter().addRow(position, matcher1.group(1), matcher1.group(2));
                            }
                        }
                    }
                }
            }
        };
        http = new HttpManager(handler);
        http.setParams(params);
        http.setUA("Mozilla/5.0 (Linux; Android 15.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Mobile Safari/537.36");
        getResult();
    }

    void getDetail(String url) {
        http.getRequest("https://tice.sysu.edu.cn" + url, 3);
    }

    void getCreditDetail(String url) {
        http.getRequest("https://tice.sysu.edu.cn" + url, 4);
    }

    void getResult() {
        http.getRequest("https://tice.sysu.edu.cn/m/tice", 0);
    }

    void getCredit() {
        http.getRequest("https://tice.sysu.edu.cn/m/kwjfList", 1);
    }

    void getSwim() {
        http.getRequest("https://tice.sysu.edu.cn/m/tice/studentSwim", 2);
    }
}

File: app/src/main/java/com/sysu/edu/academic/PhysicalFitnessTestResultActivity.kt
package com.sysu.edu.academic

import android.os.Bundle
import com.google.android.material.tabs.TabLayout
import com.google.android.material.tabs.TabLayoutMediator
import com.sysu.edu.BaseActivity
import com.sysu.edu.R
import com.sysu.edu.api.CookieManager
import com.sysu.edu.api.HttpManager
import com.sysu.edu.api.TargetUrl
import com.sysu.edu.databinding.ActivityPagerBinding
import com.sysu.edu.view.Pager2Adapter
import com.sysu.edu.view.StaggerFragment
import io.reactivex.rxjava3.android.schedulers.AndroidSchedulers
import io.reactivex.rxjava3.core.Observable
import io.reactivex.rxjava3.schedulers.Schedulers
import org.jsoup.Jsoup
import org.jsoup.nodes.Document

class PhysicalFitnessTestResultActivity : BaseActivity() {
	lateinit var http: HttpManager
	var position: Int = 0
	override fun onCreate(savedInstanceState: Bundle?) {
		super.onCreate(savedInstanceState)
		val adp = Pager2Adapter(this)
		adp.add(StaggerFragment()).add(StaggerFragment()).add(StaggerFragment())
		val binding = ActivityPagerBinding.inflate(layoutInflater).apply {
			toolbar.setTitle(R.string.physical_fitness_test_result)
			toolbar.setNavigationOnClickListener { supportFinishAfterTransition() }
			pager.adapter = adp
			TabLayoutMediator(tabLayout, pager) { tab: TabLayout.Tab?, position: Int ->
				tab?.text = mutableListOf<String?>("体测成绩", "体育积分", "游泳")[position]
			}.attach()
		}
		setContentView(binding.root)
		val cm = CookieManager(this)
		val urls = mutableListOf("m/tice", "m/kwjfList", "m/tice/studentSwim")
		val name = mutableListOf(R.string.total_score, R.string.total_gym_credit, R.string.is_pass)
		val links = mutableListOf(mutableListOf(), mutableListOf(), mutableListOf<String>())
		(0..2).forEach { i ->
			get(cm, urls[i], links[i], adp.get(i) as StaggerFragment, name[i], i)
		}
	}
	
	private fun get(
		cm: CookieManager, urls: String,
		links: MutableList<String>,
		adp: StaggerFragment,
		name: Int,
		tabIndex: Int
	) {
		config.contextUtil.disposable.add(Observable.fromCallable<Any> {
			Jsoup.connect("https://tice.sysu.edu.cn/${urls}")
				.header("Cookie", cm.toSimpleString("tice.sysu.edu.cn"))
				.userAgent("Mozilla/5.0 (Linux; Android 15.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Mobile Safari/537.36")
				.timeout(3000)
				.get()
		}
			.retry(3)
			.subscribeOn(Schedulers.io())
			.observeOn(AndroidSchedulers.mainThread())
			.subscribe {
				val doc = it as Document
				if (doc.getElementById("netid-login") != null) config.contextUtil.login(
					TargetUrl.TICE
				) {
					get(cm, urls, links, adp, name, tabIndex)
				} else doc.select("a.weui-cell.weui-cell_access")
					.forEach { element ->
						val title = element.selectFirst(".weui-cell__bd p")
							?.text()
						val score = element.selectFirst("span")
							?.text()
						val link = element.attr("href")
						links.add(link)
						adp.addSection(
							title,
							R.drawable.calendar,
							mutableListOf(getString(name)),
							mutableListOf(score)
						              )

						val sectionIndex = adp.sections.size - 1
						val twoColumnsAdapter = adp.sectionAdapter.getTwoColumnsAdapter(sectionIndex)
						val currentPos = links.size - 1
						
						twoColumnsAdapter.setRowClickListener(0) {
							if (twoColumnsAdapter.itemCount == 1 && tabIndex != 2) {
								config.contextUtil.disposable.add(Observable.fromCallable<Any> {
									Jsoup.connect("https://tice.sysu.edu.cn${links[currentPos]}")
										.header("Cookie", cm.toSimpleString("tice.sysu.edu.cn"))
										.userAgent("Mozilla/5.0 (Linux; Android 15.0; Nexus 5 Build/MRA58N) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/144.0.0.0 Mobile Safari/537.36")
										.timeout(3000)
										.get()
								}
									.retry(3)
									.subscribeOn(Schedulers.io())
									.observeOn(AndroidSchedulers.mainThread())
									.subscribe { detailDoc ->
										(detailDoc as Document).select(".weui-cell")
											.forEach { cellElement ->
												when (tabIndex) {
													0 -> {
														val cellTitle = cellElement.selectFirst(".weui-cell__bd p")?.text()
														val cellScore = cellElement.selectFirst(".weui-cell__ft")?.text()
														twoColumnsAdapter.add(key = cellTitle, value = cellScore)
													}
													1 -> {
														twoColumnsAdapter.add(
															key = cellElement.selectFirst("p")?.text() ?: "",
															value = cellElement.selectFirst(".weui-cell__hd")?.text() ?: ""
														)
														cellElement.select(".container").forEach { item ->
															twoColumnsAdapter.add(
																key = item.selectFirst(".left_side")?.text() ?: "",
																value = item.selectFirst("p")?.text() ?: ""
															)
														}
													}
												}
											}
									})
							}
						}
					}
			})
	}
	
	override fun onDestroy() {
		super.onDestroy()
		config.contextUtil.disposable.dispose()
	}
}